### ISSUE TITLE
UX trial run: starting two nodes on one machine

### ISSUE THREAD

**Issue by ZenGround0** [filecoin-project/go-filecoin#666]
### Description
Hey all, I spent a few hours today trying to get two filecoin nodes on my local network into a little swarm as a way to generate a multi-block tipset via CLI.  This was motivated to verify UX changes I've been making as part of #543.  While I finished just short of my goal I learned a lot about where we are failing for UX in the case of multiple nodes per swarm and per machine.  I'm sharing the long winded notes generated during debugging the trial run as well as a more succinct set of TODOs that will make this process easier for people doing the same thing  once they are shipped.

#### Long Winded Notes
Today we are going to try to get a multi block tipset from the CLI

(1) We need two filecoin daemons running on one machine via the CLI
         - If you naively try to start two daemons then you get the error “daemon is already running”
         - This is ErrAlreadyRunning from commands/errors.go, the one place it is returned is commands/main.go so lets check that out
	- It turns out there is a whole function that checks for this condition called `daemonRunning`, also in main.go
	- Ok so go-filecoin doesn’t use a lock file yet and just checks if the port is used.  We could get around this easily by changing the api port in the config, but then we’d have multiple writes to the repo which is probably real bad.  Idea: we should initialize another repo for another node and also change the api port in that node’s config
	- At a glance go-filecoin init does not provide options for setting the node’s repo.  Looking at IPFS I observe an IPFS_PATH environment variable.  Let’s look at the filecoin node to see if something similar work here.  
	- Score!  FIL_PATH is the variable in question.  Furthermore it is only used for init so there shouldn’t be any collisions when setting up this new node alongside the running node.
	- I’m using fish so `set -g -x FIL_PATH ~/anotherFIL`
	- hmm ./go-filecoin init does not seem to work
	- switching to bash (from fish) where things make more sense
	- :( this also fails
	- After tons of time with this I realized that the commands are being sent to the WRONG DAEMON!  Because there is another FC daemon without this env variable set that is running and by default it is being sent all the commands
	- Success! two repos
	- In retrospect the —repodir flag would probably work without problem
	- In fact it does!  so the running daemon problem only comes up if you are using the environment variable
	- Ok now changing the config in the default spot to 3454 and keeping the config in the new spot at 3453
	- Also I’m changing the swarm address to /ip4/127.0.0.1/tcp/6001 in the default repo
	- daemon already running in the new repo
	- Hmm when I try to start the second one I get “daemon already running”, even when I call with the repo dir specifically set.  Looks like I need to use the api address flag even after modifying the config?
	- Wait after initializing the repo I don’t have an “api” file, but I need one for this to work
	- creating an api file.  Nvm this was a red herring.
	- OK the problem here is that even though the config is set to the correct api address and ostensibly this should be read to set the api addr the req also carries the default “:3453” with it, so the config is never hit.  Another nit related to this is that the address variable only contains the port.
	- The culprit seems to be a lack of update since the config came into being, the `defaultAPIAddr()` function has a comment “Until we have a config file, we need an easy way to influence the API address for testing”.
	- Yup the default value is set up with the apiaddr flag and at this point that is preventing the config from being read properly.  
	- Note this probably means we can get things working by using the apiAddr flag directly.
	- OK interesting.  now when I have started one demon with the non-default repo and I try to start another daemon with the default repo I get a locking error.  1. this means that the comment about updating once there is a lock that can be checked is out of date because there IS a lock that can be checked. 2. this means that the default repo behavior isn’t working as expected.  When 1 daemon is running another new one starts to see that daemon’s repo as the current repo.  Hmmm…
	- but if I run —cmdapiaddr ;3454 —repodir ~/.filecoin daemon —mockMine then it works
(2) have them mine on their own
	- I observe that the genesis block is the same between two separately inited repos.  Whew!  that’s good.  In that case we can achieve a multi block tipset by having each of these dudes mine.  Then connect them over the network.
	- ok back to flag problems.  Right now I have two daemons running.  One at ~/.filecoin listening on 3454 and on at ~/anotherFIL listening at 3453.  I want to look at the config via the daemon.  I use command go-filecoin —repodir ~/.filecoin config, I end up getting the config from the repo at ~/anotherFIL.   WTF you may ask?  The root problem is the same mentioned above.  The cmdapiaddr flag defaults to 3453 so even if you specify the other repo the filecoin node says “sorry no thanks” SILENTLY and shows you the repo at 3453, tough cookies.  These cmdapiaddr and repodir flag inconsistencies just get worse and worse when you look at them.
        - alright moving on (making sure to use cmdapiaddrs instead of repo dirs to send commands to the proper daemon) I notice that these peers aren’t automatically connected.  Works for me right now.
        - OK calling mining once on each of them we now have two different blocks stemming from the same genesis block.  Both are in the same tipset but each node only knows about one block

(3) add these peers together into a swarm
	- OK lets see if we can connect these local peers using `filecoin swarm connect`
	- this command appears to take in an address, so I’m assuming multi address and so I’ll try to connect 3453 to the listening adds of 3454: ""
	- I get “invalid peer address, invalid IPFS Address”, ok.  Maybe I should use the peer id
	- ok, trying “/ipfs/QmY7uDVepWvsJegsF1pC37LCJNBPsLUGJkvyB1JtBDbS1C”, I get “dial attempt failed.  Maybe I need to add on the network address too.
	- success! with address: "/ip4/127.0.0.1/tcp/6001/ipfs/QmY7uDVepWvsJegsF1pC37LCJNBPsLUGJkvyB1JtBDbS1C/“, which concatenates network address to ipfs address
	- This peer now shows up in the `swarm peers list`
	- Note that this has NOT changed the heaviest tipset as seen by the chain command (which I updated to report multiple tipsets).  The logs of both nodes do not mention hearing about new blocks or processing new blocks.
	- Now I’ll connect 3454 to 3453, oh actually it is automatically connected, cool.
	
(4) getting a multi block tipset???
	- I was hoping the hello mechanism would have kicked in and reported to each node that there is a block they should process (even though hello still works with the outdated getbestblock it would work in this case because each node has only one block).  However this is not happening.  
	- There are no commands I can see to force a sync or best block propagation manually.
	- shutting down and restarting does not force block sharing
	- reconnecting after shutdown does not force block sharing
	- It looks like work to get blocks shared is still TODO, and so we can’t get a multi-block tipset this way just yet.
	- OK update, if one node mines as both are ALREADY connected, then the same chain is synced!  Hooray!  This still doesn’t solve the multi block tipset creation needs because this means one node extends the chain and the non-mining node gets left behind.  We can try to have them both mine at about the same time and get a Multiblocktipset out of a race but definitely not reliable.

#### TL;DR Output TODOS
1. We probably should remove the FIL_PATH option and just rely on the repodir flag.  Fewer options is easier to think about.  If we really want to keep FIL_PATH we should ensure that ./go-filecoin init runs commands in its own process, not against any other running daemon that doesn’t have FIL_PATH set, otherwise we open the door to a quite confusing issue to users.
2. I could see the cmdapiaddr option being useful, but again a single way of doing things is nicer IMO.  Interested to learn if others disagree strongly.  To remove this flag and rely entirely on the config we would really want to make the go-filecoin config option not rely on the daemon already running so it would be easy to do `go-filecoin config api.addr ‘ “:3454” ‘ before running the daemon so everything would work out.  Otherwise we'd have to rely on users manually editing the config.
3. If we don’t remove the flag entirely we definitely need to fix how the flag and repo cooperate.  This is also an easy interim fix if we eventually want to nuke the cmdapiaddr flag.  Right now a default flag value overrides a modified config value which != good in my book.  There is a really straightforward fix mentioned in the comments, all we do is get rid of the default on the cmdapiaddr flag and let the default encoded into the repo on inception do the work for us.
4. when checking for an existing daemon it still makes sense to look for an unavailable port, but likely the node can try another port in this case.  We should start checking for lock file contention too as suggested by the comment in `daemonRunning`.
5. I would like some way to automatically tell my node to tell everyone about its heaviest tipset for playing around like this. Nodes should also successfully update their local chain when nodes with new chain info JOIN, not just when already connected nodes start to mine.

### Acceptance criteria
We should close these issues when all TODOs are addressed to the level where setting up a 2 node local swarm avoids all of the issues I ran into.  Ideally this also includes the swarm properly sharing heaviest tipsets upon joining though maybe that belongs in another issue.

**Comment by anorth** [comment]
@ZenGround0 @travisperson is much of this issue still relevant? Has FAST solved the majority of issues, such that we might file new, tightly scoped ones for other improvements?
